nucypher/crypto/utils.py

```diff
@@ -47,39 +50,89 @@ def canonical_address_from_umbral_key(public_key: UmbralPublicKey) -> bytes:
-            raise ValueError(f"When not passing a v value, "
-                             f"the signature size should be {expected_signature_size} B.")
-    elif v_value_to_try in (0, 1, 27, 28):
-        expected_signature_size = ecdsa_signature_size
-        if not len(signature) == expected_signature_size:
-            raise ValueError(f"When passing a v value, "
-                             f"the signature size should be {expected_signature_size} B.")
+    if v_value_to_try in (0, 1, 27, 28):
         if v_value_to_try >= 27:
             v_value_to_try -= 27
         signature = signature + v_value_to_try.to_bytes(1, 'big')
     else:
         raise ValueError("Wrong v value. It should be 0, 1, 27 or 28.")
 
+    kwargs = dict(hasher=None) if is_prehashed else {}
     pubkey = PublicKey.from_signature_and_message(serialized_sig=signature,
-                                                  message=prehashed_message,
-                                                  hasher=None)
+                                                  message=message,
+                                                  **kwargs)
     return pubkey.format(compressed=True)
+
+
+def get_signature_recovery_value(message: bytes,
+                                 signature: Union[bytes, Signature],
+                                 public_key: Union[bytes, UmbralPublicKey],
+                                 is_prehashed: bool = False) -> bytes:
+    """
+    Obtains the recovery value of a standard ECDSA signature.
+
+    :param message: Signed message
+    :param signature: The signature from which the pubkey is recovered
+    :param public_key: The public key for verifying the signature
+    :param is_prehashed: True if the message is already pre-hashed. Default is False, and message will be hashed with SHA256
+    :return: The compressed byte-serialized representation of the recovered public key
+    """
+
+    signature = bytes(signature)
+    ecdsa_signature_size = Signature.expected_bytes_length()
+    if not len(signature) == ecdsa_signature_size:
+        raise ValueError(f"The signature size should be {ecdsa_signature_size} B.")
+
+    kwargs = dict(hasher=None) if is_prehashed else {}
+    for v in (0, 1):
+        v_byte = bytes([v])
+        recovered_pubkey = PublicKey.from_signature_and_message(serialized_sig=signature + v_byte,
+                                                                message=message,
+                                                                **kwargs)
+        if bytes(public_key) == recovered_pubkey.format(compressed=True):
+            return v_byte
+    else:
+        raise ValueError("Signature recovery failed. "
+                         "Either the message, the signature or the public key is not correct")
+
+
+def get_coordinates_as_bytes(point: Union[Point, UmbralPublicKey, SignatureStamp],
+                             x_coord=True,
+                             y_coord=True) -> bytes:
+
+    try:
```

Comment:
What's happening in this `try` block?
Is what meant this?
```
if isinstance(point, Point):
    ...
elif isinstance(point, UmbralPublicKey):
    ...
else:
    ...
```

Reply:
The idea is that, whatever you pass into the function (Point, UmbralPublicKey or SignatureStamp), we need its uncompressed point representation at some moment. For the first two cases you can just call `.to_bytes(is_compressed=False)`, but in case it's a stamp then you need the public key first. Some Pythonistas (like @jMyles and @KPrasch) preach that it's [easier to ask for forgiveness than permission](https://docs.python.org/2/glossary.html#term-eafp) (which in this case implies the use of `try/except` rather than `isinstance`) and I'm starting to believe it 😛

Reply:
Eh, I think in this particular case it makes the code less readable: too implicit for even humans to understand. A random reader would not be able to guess that it's SignatureStamp which has `as_umbral_pubkey` method.

Isn't this one more clear?
```
if isinstance(point, SignatureStamp):
    point = point.as_umbral_pubkey()

coordinates_as_bytes = point.to_bytes(is_compressed=False)[1:]
...
```

Reply:
Gotcha. I'll make the changes 👍

Reply:
✅